subsys/bluetooth/host/hci_core.c

```diff
@@ -4020,6 +4020,22 @@ int bt_enable(bt_ready_cb_t cb)
 	k_thread_name_set(&bt_workq.thread, "BT RX");
 #endif
 
+	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+#if defined(CONFIG_BT_HCI_SET_PUBADDR)
+		if (bt_dev.drv->set_public_address != NULL) {
+			err = bt_dev.drv->set_public_address(&bt_dev.id_addr[BT_ID_DEFAULT].a);
```

Comment:
I was originally thinking that this would happen *after* the call to `drv->open()`, since usually this operation would require having the HCI transport available (which `open()` ensures), unless this requires the driver to temporarily store the address before actually writing it to the controller. Any reason why you think this absolutely needs to be before `open()`?

Reply:
I'm testing it right now, but I think BlueNRG may require setting this option before `open()` completes.

Reply:
Right after `drv->open()` is fine actually, just not after `common_init()`. I've put it right after `drv->setup()` now.

Reply:
> Right after `drv->open()` is fine actually, just not after `common_init()`. I've put it right after `drv->setup()` now.

Could we do it right before setup() instead, in case the setup routine wants to use this address? E.g. the `hci_psoc6_bless.c` HCI driver uses the `setup()` callback to program a public address to the controller:
https://github.com/zephyrproject-rtos/zephyr/blob/063ce9caf54fa656f02ae48f3c9d537659a10dec/drivers/bluetooth/hci/hci_psoc6_bless.c#L199-L208

One could imagine that someone might want to extend support in this driver to override the default public address with another one provided from the app, in which case one way to do this would be to use the `set_public_addr()` callback to store the alternate address and then program this instead of the default one in `setup()`.

Reply:
Makes sense, done.

Reply:
> I'm testing it right now, but I think BlueNRG may require setting this option before `open()` completes.

`bt_spi_open` will reset the BlueNRG device, so everything you do before is going to be lost.

Reply:
No, I had previously sent the ACI command "in" `open()` (in `bt_spi_handle_vendor_evt()`), which worked, but then moved it after `common_init()`, which didn't work anymore. It doesn't actually have to happen in `open()`, just needs to be before `common_init()`, not entirely sure why.

Reply:
OK, I think this is because that command needs to be run after reset, since it is disabled after standard command are received. I am quoting the programming manual:

> The command ACI_HAL_WRITE_CONFIG_DATA should be sent to BlueNRG and 
> BlueNRG-MS devices by the uC before starting BLE operations (after each power-up or 
> reset of BlueNRG).

I guess you run the command in the `common_init` after some other commands. On the other hand, I would have expected the error 0x0C (Command Disallowed) returned.